### PR TITLE
chore(warehouse): handle bucket path styles

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

> Handle Path-style access and Virtual-hosted–style access for S3 location.
> Added unit tests for utils.

## Notion Ticket
https://www.notion.so/rudderstacks/Handle-bucket-name-config-for-warehouses-be1699b5312a40ea86dc8e275b649bf5

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/utils/utils_test.go [hunk 1]
```diff
@@ -1,139 +1,917 @@
...
+				"region":  "",
+			},
+		},
+	}
+	for idx, input := range inputs {
+		got, err := CaptureRegexGroup(input.regex, input.pattern)
+		assertNoError(t, err)
+		if !reflect.DeepEqual(got, input.groups) {
+			t.Errorf("got %#v want %#v input %d", got, input.groups, idx)
+		}
```
Comment (abhimanyubabbar) [inline warehouse/utils/utils_test.go:141]: There is a very light library `testtify/require` which can be used to perform these assertions. Will streamline the statements I think. Wdyt ?

File: warehouse/utils/utils_test.go [hunk 2]
```diff
@@ -1,139 +1,917 @@
...
+			s3Location: "s3://test-bucket/test-object.csv",
+			region:     "any-region",
+		},
+		{
+			location:   "https://my.test-bucket.s3.amazonaws.com/test-object.csv",
+			s3Location: "s3://my.test-bucket/test-object.csv",
+			region:     "",
+		},
+		{
+			location:   "https://my.test-bucket.s3.us-west-1.amazonaws.com/test-object.csv",
```
Comment (Jayachand) [inline warehouse/utils/utils_test.go:45]: what happens when bucket name is `my.example.s3.bucket`

File: warehouse/utils/utils.go [hunk 3]
```diff
@@ -153,6 +153,10 @@ func Init() {
 
 func loadConfig() {
 	IdentityEnabledWarehouses = []string{SNOWFLAKE, BQ}
+	s3Regexes = []string{
+		"https?://s3([.-](?P<region>[^.]+))?.amazonaws.com/(?P<bucket>[^/]+)/(?P<keyname>.*)",
+		"https?://(?P<bucket>[^/]+).s3([.-](?P<region>[^.]+))?.amazonaws.com/(?P<keyname>.*)",
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/utils/utils.go]: Done.

File: warehouse/utils/utils_test.go [hunk 4]
```diff
@@ -1,139 +1,917 @@
...
-					}
-
-					locations := GetS3Locations(inputs)
-					Expect(locations[0].Location).To(Equal("s3://test-bucket/test-object.csv"))
-					Expect(locations[1].Location).To(Equal("s3://test-bucket/test-object.csv"))
-					Expect(locations[2].Location).To(Equal("s3://my.test-bucket/test-object.csv"))
-					Expect(locations[3].Location).To(Equal("s3://my.test-bucket/test-object.csv"))
-				})
-			})
+func TestGetS3Location(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/utils/utils_test.go:21]: Added empty string test cases as well. Also added the same for TestCaptureRegexGroup as well.

File: warehouse/utils/utils_test.go (hunk 2 above)
Comment (achettyiitr) [reply warehouse/utils/utils_test.go:43]: Also added in the test suite as well. Thanks @Jayachand for the feedback.

**Review by abhimanyubabbar — APPROVED** [review]
Changes look good 👍🏼

**Review by Jayachand — APPROVED** [review]
LGTM

**Pull request merged**